dfu with sidewalk fuota

ktaborowski opened the pull request:
## CI parameters

```yaml
Github_actions:
  #(branch, hash, pull/XXX/head)
  NRF_revision: main

  # Do not change after creating PR
  Create_NRF_PR: false
Jenkins:
  test-sdk-sidewalk: master
```

## Description

JIRA ticket: 

## Self review

- [x] There is no commented code.
- [x] There are no TODO/FIXME comments without associated issue ticket.
- [x] Commits are properly organized.
- [x] Change has been tested.
- [x] Tests were updated (if applicable).

RobertGalatNordic commented on samples/sid_end_device/src/sidewalk.c:
```diff
@@ -316,31 +317,26 @@ static void state_sidewalk_run(void *o)
 		}
 		LOG_INF("Received file Id %d; buffer size %d; file offset %d", args->desc.file_id,
 			args->buffer->size, args->desc.file_offset);
-		uint8_t hash_out[32];
-		sid_pal_hash_params_t params = { .algo = SID_PAL_HASH_SHA256,
-						 .data = args->buffer->data,
-						 .data_size = args->buffer->size,
-						 .digest = hash_out,
-						 .digest_size = sizeof(hash_out) };
-
-		sid_error_t e = sid_pal_crypto_hash(&params);
-		if (e != SID_ERROR_NONE) {
-			LOG_ERR("Failed to hash received file transfer with error %s",
-				SID_ERROR_T_STR(e));
-		} else {
-#define HEX_PRINTER(a, ...) "%02X"
-#define HEX_PRINTER_ARG(a, ...) hash_out[a]
-			char hex_str[sizeof(hash_out) * 2 + 1] = { 0 };
-			snprintf(hex_str, sizeof(hex_str), LISTIFY(32, HEX_PRINTER, ()),
-				 LISTIFY(32, HEX_PRINTER_ARG, (, )));
-			LOG_INF("SHA256: %s", hex_str);
```
Our CI tests rely on the hash output of the payload so we shouldn't delete this (at liest not without changes made in tests)

 would propose to extract this to a function (to help with swapping this implementation to new one) and contitionally call the extracted function ( or the new implementation) based on the KConfig feature flag setting. 
 
 I would propose to set the default value of this feature flag to use the old implementation.
 
 Not everything need to be behind the feature flag setting. The changes in the structures, additional fields does not impact previous implementation, so they can be added as is.

ktaborowski replied:
thanks for pointing this out, will be reversed before converting draft to pr

RobertGalatNordic commented on samples/sid_end_device/src/file_transfer.c:
```diff
@@ -159,6 +201,54 @@ void app_file_transfer_demo_init(struct sid_handle *handle)
+		if (ret != SID_ERROR_NONE) {
+			LOG_ERR("sbdt cancel ret %s", SID_ERROR_T_STR(ret));
+		}
+	} else {
```
Just a cosmetic suggestion, this else is not needed

RobertGalatNordic commented:
/__w/sdk-sidewalk/sdk-sidewalk/sidewalk/samples/sid_end_device/src/sidewalk.c:180:13: error: 'print_hash' defined but not used [-Werror=unused-function]

RobertGalatNordic commented on samples/sid_end_device/src/sidewalk.c:
```diff
@@ -175,6 +177,28 @@ static void state_sidewalk_entry(void *o)
 #endif /* CONFIG_SIDEWALK_AUTO_START */
 }
 
+static void print_hash(void *data, size_t data_size)
```
unused when Kconfig CONFIG_SIDEWALK_FILE_TRANSFER disabled

ktaborowski committed:
samples: add single image dfu with Sidewalk fuota

[KRKNWK-18739]
* modify file transfer module - support mcuboot
* write image to flash on file transfer event
* add reboot event
* move scratch buffer to utils
* group mcuboot log configs in prj.conf
* sort out heders, fix formating etc

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

samples: fix flie transfer testing

tbd

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

samples: disable strem flash progress

See https://github.com/nrfconnect/sdk-nrf/pull/14684

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

samples: fix build error

tbd

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

ktaborowski committed:
samples: file transfer improvements

[KRKNWK-18739]
* init dfu image on every new transfer
* file transfer in default hello app
* fix build issues

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

samples: minor fix

tbd

ktaborowski committed:
samples: add experimental sidewalk queue timeout

[KRKNWK-19101]
* wait to put msg if not in isr
* increase queue size

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

ktaborowski committed:
utils: add dfu mutli image utils and refactor

[KRKNWK-18739]
* separate nrf img management modult to utils
* refactor utils (move on_dev_cert to subsys)

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

tests: fix unit tests

tbd

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

ktaborowski committed:
doc: describe dfu options in sidewlak sample

[KRKNWK-18739]

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

totyz reviewed (approved):
Tests PASSED

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+The sample supports over-the-air (OTA) device firmware upgrade (DFU) using one of the two following protocols:
+
+* Simple Management Protocol (SMP) over Bluetooth® LE.
+   In this case, the DFU can be done either using a smartphone application or a PC command line tool.
+   Note that this protocol is not part of the Sidewalk specification.
```
```suggestion
  In this case, the DFU is performed using a smartphone application or a PC command line tool.
   This protocol is not a part of the Sidewalk specification.
```

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+   In this case, the DFU can be done either using a smartphone application or a PC command line tool.
+   Note that this protocol is not part of the Sidewalk specification.
+* Downlink file transfer over Bluetooth LE (experimental).
+   The Sidewalk Bulk Data Transfer mode and Integration of AWS IoT FUOTA service allow sending files to fleet of IoT devices from the AWS IoT FUOTA task.
```
```suggestion
  This method uses the Sidewalk Bulk Data Transfer (SBDT) mode and integration of AWS IoT FUOTA service to send files to fleet of IoT devices from the AWS IoT FUOTA task.
```

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+   Note that this protocol is not part of the Sidewalk specification.
+* Downlink file transfer over Bluetooth LE (experimental).
+   The Sidewalk Bulk Data Transfer mode and Integration of AWS IoT FUOTA service allow sending files to fleet of IoT devices from the AWS IoT FUOTA task.
+   For more details, see Sidewalk Bulk Data Transfer section in the `Amazon Sidewalk specification`_.
```
```suggestion
   For more details, see the Sidewalk Bulk Data Transfer section in the `Amazon Sidewalk specification`_.
```

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+
+   .. note::
+
+      To test this functionality, it is required to have access to S3 bucket.
+      If you want to gain the access, contact your account manager or sidewalk-beta-support@amazon.com.
+      In the message provide your AWS account ID to be added to the allowlist.
```
```suggestion
      To test this functionality, you must have access to S3 bucket.
      To gain it, contact your account manager or sidewalk-beta-support@amazon.com.
      In the message provide your AWS account ID to be added to the allowlist.
```

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+      If you want to gain the access, contact your account manager or sidewalk-beta-support@amazon.com.
+      In the message provide your AWS account ID to be added to the allowlist.
+
+In both cases, MCUboot secure bootloader is used to apply the new firmware image.
```
```suggestion
In both cases, the MCUboot secure bootloader is used to apply the new firmware image.
```

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -129,6 +129,32 @@ For example:
+
+The following configuration arguments are available during the build process for configuring DFU:
+
+* To configure the sample to support the downlink file transfer over Bluetooth LE, use the ``-DCONFIG_SIDEWALK_FILE_TRANSFER=y`` build flag.
```
```suggestion
* To enable support for the downlink file transfer over Bluetooth LE, use the ``-DCONFIG_SIDEWALK_FILE_TRANSFER=y`` build flag.
```

ktaborowski committed:
doc: update memory requirements

Update sidewalk samples memory requrements table

Signed-off-by: Krzysztof Taborowski <krzysztof.taborowski@nordicsemi.no>

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -135,25 +135,25 @@ Device Firmware Upgrade support
-   In this case, the DFU can be done either using a smartphone application or a PC command line tool.
-   Note that this protocol is not part of the Sidewalk specification.
+  In this case, the DFU is performed using a smartphone application or a PC command line tool.
+   This protocol is not a part of the Sidewalk specification.
```
```suggestion
  This protocol is not a part of the Sidewalk specification.
```
sorry - forgot to align last sentences of these 2 bullet points

annwoj commented on doc/samples/sid_end_device.rst:
```diff
@@ -135,25 +135,25 @@ Device Firmware Upgrade support
-   The Sidewalk Bulk Data Transfer mode and Integration of AWS IoT FUOTA service allow sending files to fleet of IoT devices from the AWS IoT FUOTA task.
-   For more details, see Sidewalk Bulk Data Transfer section in the `Amazon Sidewalk specification`_.
+  This method uses the Sidewalk Bulk Data Transfer (SBDT) mode and integration of AWS IoT FUOTA service to send files to fleet of IoT devices from the AWS IoT FUOTA task.
+   For more details, see the Sidewalk Bulk Data Transfer section in the `Amazon Sidewalk specification`_.
```
```suggestion
  For more details, see the Sidewalk Bulk Data Transfer section in the `Amazon Sidewalk specification`_.
```

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
       +----------------------------------------------------------------------------------------------------------------+------------------+----------------------+-------------------+--------------------------+----------------+----------------+
-      |  :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                               |      26.33       |        496.32        |         4         |            8             |     534.65     |     109.91     |
+      |                :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                 |      26.33       |        512.94        |         4         |            8             |     551.27     |     122.16     |
       +----------------------------------------------------------------------------------------------------------------+------------------+----------------------+-------------------+--------------------------+----------------+----------------+
-      |  :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Release)                             |      26.33       |        430.45        |         4         |            8             |     468.78     |     105.04     |
+      |               :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Release)                |      26.33       |        430.49        |         4         |            8             |     468.82     |     106.04     |
```
can these empty spaces be aligned as per the previous version?

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
 
       The following table lists the total memory requirements for the default variant of the :ref:`Sidewalk_End_device` sample running on the `nRF5340 DK`_.
-      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
+      To determine the available space for user extension, it needs to be summarized versus the allocated partition size and its layout.
```
```suggestion
      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
```
as per previous version - in technical documentation we avoid using latin words

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
+      |  :ref:`Hello Sidewalk Bluetooth LE <variant_sidewalk_hello>` (``CONFIG_SIDEWALK_SUBGHZ_SUPPORT=n``, Release)   |      30.77       |        291.82        |         4         |            8             |     334.59     |     79.40      |
       +----------------------------------------------------------------------------------------------------------------+------------------+----------------------+-------------------+--------------------------+----------------+----------------+
-      |  :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                               |      30.77       |        435.18        |         4         |            8             |     477.95     |     98.34      |
+      |                :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                 |      30.77       |        452.05        |         4         |            8             |     494.82     |     110.62     |
```
as above

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
 
       The following table lists the total memory requirements for the default variant of the :ref:`Sidewalk_End_device` sample running on the `nRF54L15 PDK`_.
-      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
+      To determine the available space for user extension, it needs to be summarized versus the allocated partition size and its layout.
```
```suggestion
      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
```

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
+      |  :ref:`Hello Sidewalk Bluetooth LE <variant_sidewalk_hello>` (``CONFIG_SIDEWALK_SUBGHZ_SUPPORT=n``, Release)   |      26.15       |        285.78        |         4         |            8             |     323.93     |     84.77      |
       +----------------------------------------------------------------------------------------------------------------+------------------+----------------------+-------------------+--------------------------+----------------+----------------+
-      |  :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                               |      26.03       |        432.32        |         4         |            8             |     470.35     |     103.99     |
+      |                :ref:`Hello Sidewalk Bluetooth LE and sub-GHz <variant_sidewalk_hello>` (Debug)                 |      26.15       |        449.00        |         4         |            8             |     487.15     |     116.28     |
```
as above

annwoj commented on doc/includes/memory_requirement.txt:
```diff
@@ -3,50 +3,50 @@
 
       The following table lists the total memory requirements for the default variant of the :ref:`Sidewalk_End_device` sample running on the `nRF52840 DK`_.
-      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
+      To determine the available space for user extension, it needs to be summarized versus the allocated partition size and its layout.
```
```suggestion
      To determine the available space for user extension, it needs to be summarized against the allocated partition size and its layout.
```
